### PR TITLE
Lookup params @hlandau's approach

### PR DESCRIPTION
**Pull request by paulidale**
As per  #20888 but with @halndau's alternative suggestion implemented for comparison (just the final commit).

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: include/internal/param_names.h.in
Comment (levitte) [inline include/internal/param_names.h.in:15]: Couldn't this `NUM_PIDX`?

Comment (levitte) [reply include/internal/param_names.h.in:15]: ... that way, the array boundary wouldn't be much of a worry, right? (we could still do with an assertion to make sure we don't do stupid things, but yeah)

**Comment by levitte** [comment]
I recall you say that in the end, it comes down to choosing between a `memset()` call when scanning the keys, or a `for-switch` loop around the result.

Yup, I see what you mean.  It took me a few reads...

File: include/internal/param_names.h.in
Comment (paulidale) [reply include/internal/param_names.h.in:15]: Some parameters can be repeated and concatenated, so no finite limit is perfect.

Comment (levitte) [reply include/internal/param_names.h.in:15]: Gah, you're right, I forgot about that.

File: providers/implementations/ciphers/ciphercommon_gcm.c
Comment (hlandau) [inline providers/implementations/ciphers/ciphercommon_gcm.c:155]: It's good to be paranoid, but in this case I'd rather rely on our invariant above so we can elide this. Since this structure would be adopted uniformly throughout countless functions, I think conciseness is more valuable here, and we can ensure this invariant holds in our central functions anyway.

File: include/internal/param_names.h.in
Comment (hlandau) [inline include/internal/param_names.h.in:20]: I was kind of thinking of having this array length be specified at each function using these functions, so that it can be long enough as needed for that specific function. It'd be similar to the pattern we often use to construct OSSL_PARAMs in the first place.

However, I guess the issue is the OSSL_PARAM semantics are to ignore unrecognised parameters. So theoretically a client could stuff in an infinite number of irrelevant parameters. That's certainly a problem.

How would the original PR handle multiple instances of the same parameter, for that matter? I assume callers aren't required to pass multiple instances of the same parameter consecutively. Hmm...

**Comment by hlandau** [comment]
>However, I guess the issue is the OSSL_PARAM semantics are to ignore unrecognised parameters. So theoretically a client could stuff in an infinite number of irrelevant parameters. That's certainly a problem.

Actually, thinking of this. If we did adopt codegen to make bespoke trie or perfect hashers for each user of OSSL_PARAM in our in-house providers, I guess in principle we could just have those generated functions not stash in an array for parameters we don't intend to support, and even enforce how many of each parameter may be provided. So this would allow the array length to be bounded on a per-function basis.

**Comment by paulidale** [comment]
> If we did adopt codegen to make bespoke trie or perfect hashers for each user of OSSL_PARAM in our in-house providers, I guess in principle we could just have those generated functions not stash in an array for parameters we don't intend to support, and even enforce how many of each parameter may be provided. So this would allow the array length to be bounded on a per-function basis.

Exactly.  That's the biggest stumbling point I see with this approach.  It can't be guaranteed to work without significant cost.  The other two approaches can & I can't see the savings here justifying themselves.

The approach in #20940 is going to be a **lot** of work to implement but it must be the fastest, safest and lowest RAM use (but not code size).

**Comment by paulidale** [comment]
Some performance results are in (AMD Ryzen 9, 10⁶ iterations):

| # params | TRIE | TRIE + this | &Delta;% |
| :--: | -----: | ------: | :-: |
| 1 of 7 | 71,231,760 | 17,761,922 | -75% |
| 7 of 7 | 146,584,164 | 94,982,444 | -35% |

This approach is faster than a TRIE + single array for few parameters present.
The downside, is that this requires code refactoring and special treatment of repeated parameters.

[benchmark.c.txt](https://github.com/openssl/openssl/files/11482755/benchmark.c.txt)

**Comment by paulidale** [comment]
Re-spun this to be closer to the final idea.  Performance doesn't appear to be altered.

**Comment by paulidale** [comment]
Reviews please?

**Comment by mattcaswell** [comment]
> Some performance results are in (AMD Ryzen 9, 10⁶ iterations):

Do we have performance results comparing this against current master param lookup performance?

**Comment by paulidale** [comment]
I didn't take tangible measurements (it's not a fair comparison IMO), but did do something less formally:

- For one param passed and queried for, master was fastest: that's expected, it's a single `strcmp(3)` vs everything else.
- For all seven passed and queried for, master was the worst (including being worse than the hash in #20965).
- I didn't take formal measurements for passing one but testing for all seven which is the problematic case being encountered by people.  The reason for this is that it makes no difference for the alternatives I investigated -- they retain the same performance since they are using a switch once per passed param (I did verify this with the TRIE & hash).  Master will still be worst because it's using multiple `strcmp(3)` calls instead.

**Review by hlandau — APPROVED** [review]
This looks good to me.

Do we want per-function trie generation, or is that overkill?

**Review by hlandau — CHANGES_REQUESTED** [review]
Actually just one nit also: the generated code (`ossl_param_find_pidx`) only uses a single space of indentation, rather than 4

**Comment by paulidale** [comment]
> Actually just one nit also: the generated code (`ossl_param_find_pidx`) only uses a single space of indentation, rather than 4

Deliberately: if I used 4, I'd have been nitted for overlong lines :smirk:

Since it's machine generated and close to unreadable to humans, is this an issue?

**Comment by paulidale** [comment]
> Do we want per-function trie generation, or is that overkill?

I'd like to investigate this option further.  The real benefit I see from per function lookup is automation.
Tedious boilerplate code should be automated to avoid both the tedium and errors that always appear.

I think it would be a lot of effort to convert our params code to such automation so I don't view it as something critical.

**Comment by paulidale** [comment]
@mattcaswell, running the third case in my earlier comment (search 1 of 7 params) for master gives a raw result of 95,339,302.

Compared to the base TRIE approach this represents a -25% difference.
Compared with the solution here, it is a -81% difference.

**Review by hlandau — APPROVED** [review]

**Comment by hlandau** [comment]
CI is having network issues.

**Comment by paulidale** [comment]
Ping for review?

**Comment by paulidale** [comment]
Rebase and squash - zero code changes.
No rereview required.

**Comment by paulidale** [comment]
Ping for @openssl/committers review please

File: util/perl/OpenSSL/paramnames.pm
Comment (mattcaswell) [inline util/perl/OpenSSL/paramnames.pm:597]: This second strcmp seems odd. I'm not sure why we want this. This seems like a new feature (we don't do this at the moment right?...or do we?)

File: util/perl/OpenSSL/paramnames.pm
Comment (mattcaswell) [inline util/perl/OpenSSL/paramnames.pm:586]: IMO, I think the indent should be as per our code standards. Don't worry about long lines.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
This looks very nice.

I assume the plan is to convert more code to use this once this is merged?

**Comment by mattcaswell** [comment]
Can we also add the generated files to .gitignore?

File: util/perl/OpenSSL/paramnames.pm
Comment (paulidale) [reply util/perl/OpenSSL/paramnames.pm:579]: Not really.

File: util/perl/OpenSSL/paramnames.pm
Comment (paulidale) [reply util/perl/OpenSSL/paramnames.pm:554]: Counters -- not essential but kind of handy to judge the benefits.

File: util/perl/OpenSSL/paramnames.pm
Comment (paulidale) [reply util/perl/OpenSSL/paramnames.pm:597]: We don't use this currently.  It's used if we're being case insensitive: `_` and `-` matching as identical.

File: util/perl/OpenSSL/paramnames.pm
Comment (paulidale) [reply util/perl/OpenSSL/paramnames.pm:586]: The generated code is almost totally unreadable to a human, I just don't see the point of _proper_ indentation.

**Comment by paulidale** [comment]
> I assume the plan is to convert more code to use this once this is merged?

I think it will depend on how #20940 develops.

My suspicion is that #20940 is too ambitious and time consuming and in that case this approach would be expanded after merge.  There is a task to identify promising candidates for conversion regardless.

**Comment by hlandau** [comment]
Personally I don't find the generated code hard to read at all, so would prefer proper indentation. But I don't mind either way.

CI is relevant.

**Review by hlandau — APPROVED** [review]

**Comment by paulidale** [comment]
It's indented now.  More pain that it should have been and still no easier to read IMO.
Rectangular selections works nicely to match the indenting.

File: providers/implementations/kdfs/hkdf.c
Comment (mattcaswell) [reply providers/implementations/kdfs/hkdf.c:289]: Sorry...I'm still missing something here.

> For these, the param can be repeated and is concatenated. 

Where does the processing of that repetition happen? There is no loop inside `ossl_param_get1_concat_octet_string` or in this function?

> it is a "do this one" flag. It's set inside the function.

That much I understand. But it immediately goes out of scope and so is never used again for a subsequent instance of `OSSL_KDF_PARAM_INFO`???

Comment (paulidale) [reply providers/implementations/kdfs/hkdf.c:289]: Here:

https://github.com/openssl/openssl/blob/2c500578fc68871eca7fabc7ee36b4027891671b/crypto/params.c#L1536-L1565


It becomes relevant once these are converted to use a loop-swtich structure which hasn't been done yet.
I expect these will be converted but the exact list hasn't been identified yet.

Comment (mattcaswell) [reply providers/implementations/kdfs/hkdf.c:289]: Ok...thanks for pointing out the loop. But that still doesn't really explain the need for the `seen_info`. Currently it appears completely unused.

> It becomes relevant once these are converted to use a loop-swtich structure which hasn't been done yet.

In which case this so-far-unused change seems premature. Let's not include it now but when we actually need to use it. It appears quite separate to the rest of this PR.

Comment (paulidale) [reply providers/implementations/kdfs/hkdf.c:289]: As soon as any of these are converted to the alternative:
```
for (p in params)
    switch (p->key)
```
this will be required because the latter params (with the same key) will double trigger.

I can convert.  I think they'd make the cut for desirability.

**Comment by paulidale** [comment]
Updated, dropping the KDF change.

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
@hlandau - please reconfirm

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
The CI is relevant.

**Comment by t8m** [comment]
Should the generated params code be generated as part of `make update` similar to the generated errors code?

**Comment by paulidale** [comment]
CI isn't relevant, at least according to @levitte when I asked on Monday.
MacOS is failing due to case sensitive file system, Windows builds are misconfigured currently.

**Comment by paulidale** [comment]
> Should the generated params code be generated as part of `make update` similar to the generated errors code?

I don't think so.  The generation takes a fraction of a second to run and doesn't require scanning the file system.  I.e. it isn't an impediment to build times doing it this way and it's easy and more flexible due to having real make dependencies.

**Comment by paulidale** [comment]
It's well beyond my knowledge about what's going awry.  I didn't touch anything remotely related to the errors that I'm aware of.  @levitte?

**Comment by mattcaswell** [comment]
This build fails in something related to param_idx.c:

https://github.com/openssl/openssl/actions/runs/5065297982/jobs/9093731739?pr=20935

**Comment by paulidale** [comment]
That's the case sensitive one @levitte mentioned.  config and Config both being used.

Make is trying to create param_idx.c by running a perl script & failing.
The script I added doesn't touch or use the config code so why it causes the failure eludes my limited understanding of the build system.

File: crypto/params_idx.c.in
Comment (t8m) [inline crypto/params_idx.c.in:12]: @paulidale what about moving this use statement to util/dofile.pl? Maybe it could help.

File: include/internal/param_names.h.in
Comment (t8m) [inline include/internal/param_names.h.in:12]: Similarly here and below as above

File: crypto/params_idx.c.in
Comment (t8m) [reply crypto/params_idx.c.in:12]: Or actually IMO this should not be needed at all? But you just need to fully specify the `paramnames::produce_decoder` below.

File: build.info
Comment (t8m) [inline build.info:80]: Or, third idea. Remove this dependency.

Comment (paulidale) [reply build.info:80]: Absolutely not.  We lack dependencies like this all over the place which makes development more error prone.
In this case, the paramname.pm script contains the list of all OSSL_PARAM names.  Not being dependent on this makes adding a name more problematic.

File: crypto/params_idx.c.in
Comment (paulidale) [reply crypto/params_idx.c.in:12]: copy/paste programming :)  I don't know what I'm doing.

File: build.info
Comment (t8m) [reply build.info:80]: Then just try it as a workaround. I cannot see what else could cause the Windows and MacOS build breakages.

Of course real fix then will have to involve build system fixes.

Comment (paulidale) [reply build.info]: Removing the dependency addresses the build issue.  Nice catch.

Comment (t8m) [reply build.info]: @levitte any idea how to make the dependency to work?

Comment (levitte) [reply build.info]: Well yeahok...  but this isn't the actual *cause* of the detected problem.  Basically, this is treating the symptom, not the actual problem.

Comment (paulidale) [reply build.info]: Agreed, it the symptom.

**Comment by paulidale** [comment]
Merged.  I'll raise an issue for getting the dependency correct.

**Comment by bernd-edlinger** [comment]
Looks like this breaks on some old perl version:
```
$ ./config 
Configuring OpenSSL version 3.2.0-dev for target linux-x86_64
Using os-specific seed configuration
Created configdata.pm
Running configdata.pm
Created Makefile.in
Created Makefile
Created include/openssl/configuration.h

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL.md file first)      ***
***                                                                ***
**********************************************************************
$ make
/usr/bin/perl "-I." "-Iutil/perl" "-Mconfigdata" "-MOpenSSL::paramnames" "util/dofile.pl" "-oMakefile" crypto/params_idx.c.in > crypto/params_idx.c
"my" variable $trieref masks earlier declaration in same scope at /home/ed/ossl/openssl/util/perl/OpenSSL/paramnames.pm line 640.
syntax error at /home/ed/ossl/openssl/util/perl/OpenSSL/paramnames.pm line 582, near "$cursor{"
Global symbol "$cursor" requires explicit package name at /home/ed/ossl/openssl/util/perl/OpenSSL/paramnames.pm line 584.
Global symbol "%trie" requires explicit package name at /home/ed/ossl/openssl/util/perl/OpenSSL/paramnames.pm line 588.
syntax error at /home/ed/ossl/openssl/util/perl/OpenSSL/paramnames.pm line 589, near "}"
syntax error at /home/ed/ossl/openssl/util/perl/OpenSSL/paramnames.pm line 636, near "}"
syntax error at /home/ed/ossl/openssl/util/perl/OpenSSL/paramnames.pm line 662, near "}"
syntax error at /home/ed/ossl/openssl/util/perl/OpenSSL/paramnames.pm line 673, near "}"
Compilation failed in require.
BEGIN failed--compilation aborted.
make: *** [crypto/params_idx.c] Error 255
$ perl -version

This is perl 5, version 18, subversion 2 (v5.18.2) built for x86_64-linux-gnu-thread-multi
(with 55 registered patches, see perl -V for more detail)

Copyright 1987-2013, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

$ 
```

**Comment by bernd-edlinger** [comment]
... but according to this statement
https://github.com/openssl/openssl/blob/8a2e74d0536c91585fbe789e0ab7b06cab0289c2/NOTES-PERL.md?plain=1#L28-L30
perl versions up to 5.10.0 should be supported

**Comment by t8m** [comment]
@bernd-edlinger please open a new issue.

**Comment by bernd-edlinger** [comment]
OK, done: #21172

**Comment by catskul** [comment]
> Merged. I'll raise an issue for getting the dependency correct.

I see this PR was closed and commented as merged, but having a hard time tracking down where it was merged. Any breadcrumb that can be added here for those arriving from other PRs attempting to see if/when related problems may have landed?

**Comment by t8m** [comment]
> > Merged. I'll raise an issue for getting the dependency correct.
> 
> I see this PR was closed and commented as merged, but having a hard time tracking down where it was merged. Any breadcrumb that can be added here for those arriving from other PRs attempting to see if/when related problems may have landed?

Please see the entries "[openssl-machine](https://github.com/openssl-machine) pushed a commit that referenced this pull request -n Jun 2, 2023"

**Comment by t8m** [comment]
It was pushed to master branch prior to the 3.2.0 release -> it is present on 3.2.x and above versions.

**Pull request closed without merging**
